base/utf32.jl

```diff
@@ -5,25 +5,239 @@ next(s::UTF32String, i::Int) = (s.data[i], i+1)
+end
+
+"
+Converts a `UTF8String` to a `UTF32String`
+
+### Returns:
+*   `::UTF32String`
+
+### Throws:
+*   `UnicodeError`
+"
+function convert(::Type{UTF32String}, str::UTF8String)
+    dat = str.data
+    # handle zero length string quickly
+    sizeof(dat) == 0 && return empty_utf32
+    # Validate UTF-8 encoding, and get number of words to create
+    len, flags = unsafe_checkstring(dat)
+    # Optimize case where no characters > 0x7f
+    flags == 0 && @inbounds return fast_utf_copy(UTF32String, Char, len, dat, true)
+    # has multi-byte UTF-8 sequences
+    buf = Vector{Char}(len+1)
+    @inbounds buf[len+1] = 0 # NULL termination
+    local ch::UInt32, surr::UInt32
+    out = 0
+    pos = 0
+    @inbounds while out < len
+        ch = dat[pos += 1]
+        # Handle ASCII characters
+        if ch <= 0x7f
+            buf[out += 1] = ch
+        # Handle range 0x80-0x7ff
+        elseif ch < 0xe0
+            buf[out += 1] = ((ch & 0x1f) << 6) | (dat[pos += 1] & 0x3f)
+        # Handle range 0x800-0xffff
+        elseif ch < 0xf0
+            pos += 2
+            ch = get_utf8_3byte(dat, pos, ch)
+            # Handle surrogate pairs (should have been encoded in 4 bytes)
+            if is_surrogate_lead(ch)
+                # Build up 32-bit character from ch and trailing surrogate in next 3 bytes
+                pos += 3
+                surr = ((UInt32(dat[pos-2] & 0xf) << 12)
+                        | (UInt32(dat[pos-1] & 0x3f) << 6)
+                        | (dat[pos] & 0x3f))
+                ch = get_supplementary(ch, surr)
+            end
+            buf[out += 1] = ch
+        # Handle range 0x10000-0x10ffff
+        else
+            pos += 3
+            buf[out += 1] = get_utf8_4byte(dat, pos, ch)
+        end
+    end
+    UTF32String(buf)
+end
+
+"
+Converts a `UTF16String` to `UTF32String`
+
+### Returns:
+*   `::UTF32String`
+
+### Throws:
+*   `UnicodeError`
+"
+function convert(::Type{UTF32String}, str::UTF16String)
+    dat = str.data
+    len = sizeof(dat)
+    # handle zero length string quickly (account for trailing \0)
+    len <= 2 && return empty_utf32
+    # get number of words to create
+    len, flags, num4byte = unsafe_checkstring(dat, 1, len>>>1)
+    # No surrogate pairs, do optimized copy
+    (flags & UTF_UNICODE4) == 0 && @inbounds return UTF32String(copy!(Vector{Char}(len), dat))
+    local ch::UInt32
+    buf = Vector{Char}(len)
+    out = 0
+    pos = 0
+    @inbounds while out < len
+        ch = dat[pos += 1]
+        # check for surrogate pair
+        if is_surrogate_lead(ch) ; ch = get_supplementary(ch, dat[pos += 1]) ; end
+        buf[out += 1] = ch
     end
-    a[end] = Char(0) # NULL terminate
-    UTF32String(a)
+    UTF32String(buf)
+end
+
+"
+Converts a UTF-32 encoded vector of `UInt32` to a `UTF16String`
+
+### Returns:
+*   `::UTF16String`
+
+### Throws:
+*   `UnicodeError`
+"
+function convert(::Type{UTF16String}, dat::Vector{UInt32})
```

Comment:
same here - there should not be conversions between vectors of integers and string types that have different bit widths